File: Part8/Fortran/FunctionConcepts/FunctionConcepts_8.f90
! Alternate implementation of get_pizza_cost

! Returns the total price of a pizza based on size, toppings and delivery.
! Size_Par: a string as either small, medium or large.
! Topping_Count_Par: integer 0+
! Delivery_Requested_Par: Delivery has been requested if true.
! Return: Total pizza cost as a real.
program main
    real, parameter :: TOPPING_PRICE = 1.50
    real, parameter :: DELIVERY_FEE = 3.00
    real, parameter :: SIZE_COST = 3.00
    
    real :: Total_Price
    
    Total_Price = get_pizza_cost ( "small", 2, .true. )
    print *, "Total pizza price is:", Total_Price
    
contains
    function get_pizza_cost ( Size_Par, Topping_Count_Par, Delivery_Requested_Par ) result ( Total_Price )
        character(len=*), intent(in) :: Size_Par
        integer, intent(in) :: Topping_Count_Par
        logical, intent(in) :: Delivery_Requested_Par
        real :: Total_Price
        
        real :: Base_Price = 12.99
        real :: Delivery_Fee = 0.0
        real :: Topping_Cost
        
        if ( Size_Par == "medium" ) then
            Base_Price = Base_Price + SIZE_COST
        end if
        
        if ( Size_Par == "large" ) then
            Base_Price = Base_Price + SIZE_COST
        end if
        
        Topping_Cost = real ( Topping_Count_Par ) * TOPPING_PRICE
        
        if ( Delivery_Requested_Par .eqv. .true. ) then
            Delivery_Fee = DELIVERY_FEE
        end if
        
        Total_Price = Base_Price + Topping_Cost + Delivery_Fee
    end function get_pizza_cost
end program main

File: Part8/Rust/FunctionConcepts/FunctionConcepts_8.rs
// Alternate implementation of get_pizza_cost

// Returns the total price of a pizza based on size, toppings and delivery.
// size_par: a string as either small, medium or large.
// topping_count_par: integer 0+
// delivery_requested_par: Delivery has been requested if true.
// Return: Total pizza cost as a f64.

const TOPPING_PRICE: f64 = 1.50;
const DELIVERY_FEE: f64 = 3.00;
const SIZE_COST: f64 = 3.00;

fn get_pizza_cost ( size_par: &str, topping_count_par: i32, delivery_requested_par: bool ) -> f64
{
    let mut total_price = 0.0;
    
    let mut base_price = 12.99;
    
    if size_par == "medium"
    {
        base_price = base_price + SIZE_COST;
    }
    if size_par == "large"
    {
        base_price = base_price + SIZE_COST;
    }
    
    let mut delivery_fee = 0.0;
    let topping_cost = topping_count_par as f64 * TOPPING_PRICE;
    if delivery_requested_par == true
    {
        delivery_fee = DELIVERY_FEE;
    }
    
    total_price = base_price + topping_cost + delivery_fee;
    
    return total_price;
}

fn main ()
{
    let total_price = get_pizza_cost ( "small", 2, true );
    println! ( "Total pizza price is: {:.2}", total_price );
}
